lib/quby/questionnaires/entities/questions/date_question.rb

```diff
@@ -3,57 +3,79 @@ module Questionnaires
       module Questions
         class DateQuestion < Question
+          POSSIBLE_COMPONENTS = %i( day month year hour minute )
+          COMPONENT_KEYS = Hash[POSSIBLE_COMPONENTS.zip %w( dd mm yyyy hh ii)]
+          COMPONENT_PLACEHOLDERS = Hash[POSSIBLE_COMPONENTS.zip %w( DD MM YYYY hh mm)]
+          DEFAULT_COMPONENTS  = %i( day month year )
+
           # For optionally giving year, month and day fields of dates their own keys
-          attr_accessor :year_key
-          attr_accessor :month_key
-          attr_accessor :day_key
+          POSSIBLE_COMPONENTS.each do |component|
+            attr_accessor "#{component}_key".to_sym
+          end
+
+          attr_accessor :components
 
           def initialize(key, options = {})
             super
 
-            @year_key  = options[:year_key].andand.to_s
-            @month_key = options[:month_key].andand.to_s
-            @day_key   = options[:day_key].andand.to_s
+            @components = options[:components] || DEFAULT_COMPONENTS
+
+            components.each do |component|
+              instance_variable_set("@#{component}_key", options[:"#{component}_key"])
+            end
+
+            add_date_validation(options[:error_explanation])
           end
 
-          def year_key
-            (@year_key || "#{key}_yyyy").to_s
+          def add_date_validation(explanation)
+            @validations << {type: :valid_date,
+                             explanation: (explanation ||
+                                           "Vul een geldige datum in (#{expected_format})")}
           end
 
-          def month_key
-            (@month_key || "#{key}_mm").to_s
+          def expected_format
+            case components.sort
+            when [:day, :month, :year]
+              'DD-MM-YYYY'
+            when [:month, :year]
+              'MM-YYYY'
+            when [:hour, :minute]
+              'hh:mm'
+            end
           end
 
-          def day_key
-            (@day_key || "#{key}_dd").to_s
+          COMPONENT_KEYS.each do |component, name|
+            define_method("#{component}_key") do
+              (instance_variable_get("@#{component}_key") || "#{key}_#{name}").to_s
+            end
           end
 
           def claimed_keys
-            [key, day_key.to_sym, month_key.to_sym, year_key.to_sym]
+            [key] + answer_keys
           end
 
           def answer_keys
-            [day_key.to_sym, month_key.to_sym, year_key.to_sym]
+            components.map do |component|
+              send("#{component}_key").to_sym
+            end
           end
 
           def as_json(options = {})
-            super.merge(year_key: year_key, month_key: month_key, day_key: day_key)
+            component_keys = components.each_with_object({}) do |component, hash|
+              hash["#{component}_key"] = send("#{component}_key")
+            end
+            super.merge(component_keys)
           end
 
           def to_codebook(questionnaire, opts = {})
```

Comment:
eigenlijk wil je altijd eerst de specs schrijven, om de te checken of de functie doet wat je denkt dat hij doet.

Maar codeboeken zijn niet ons meest gebruikte produkt, dus kan wel even wachten. 

Maar als deze code vanacht alle data_exports doet falen, wordt dat natuurlijk wel trakteren.